pkg/executor/sortexec/topn.go

```diff
@@ -166,76 +268,358 @@ func (e *TopNExec) loadChunksUntilTotalLimit(ctx context.Context) error {
+	if err != nil {
+		return err
 	}
 	return nil
 }
 
-// doCompaction rebuild the chunks and row pointers to release memory.
-// If we don't do compaction, in a extreme case like the child data is already ascending sorted
-// but we want descending top N, then we will keep all data in memory.
-// But if data is distributed randomly, this function will be called log(n) times.
-func (e *TopNExec) doCompaction(chkHeap *topNChunkHeap) error {
-	newRowChunks := chunk.NewList(exec.RetTypes(e), e.InitCap(), e.MaxChunkSize())
-	newRowPtrs := make([]chunk.RowPtr, 0, chkHeap.rowChunks.Len())
-	for _, rowPtr := range chkHeap.rowPtrs {
-		newRowPtr := newRowChunks.AppendRow(chkHeap.rowChunks.GetRow(rowPtr))
-		newRowPtrs = append(newRowPtrs, newRowPtr)
+func (e *TopNExec) executeTopNWhenSpillTriggered(ctx context.Context) error {
+	// idx need to be set to 0 as we need to spill all data
+	e.chkHeap.idx = 0
+	err := e.spillTopNExecHeap()
+	if err != nil {
+		return err
+	}
+
+	// Wait for the finish of chunk fetcher
+	fetcherWaiter := util.WaitGroupWrapper{}
+	// Wait for the finish of all workers
+	workersWaiter := util.WaitGroupWrapper{}
+
+	// Fetch chunks from child and put chunks into chunkChannel
+	fetcherWaiter.Run(func() {
+		e.fetchChunksFromChild(ctx)
+	})
+
+	for i := range e.spillHelper.workers {
+		worker := e.spillHelper.workers[i]
+		workersWaiter.Run(func() {
+			worker.run()
+		})
+	}
+
+	fetcherWaiter.Wait()
+	workersWaiter.Wait()
+	return nil
+}
+
+func (e *TopNExec) executeTopN(ctx context.Context) {
+	defer func() {
+		if r := recover(); r != nil {
+			processPanicAndLog(e.resultChannel, r)
+		}
+
+		close(e.resultChannel)
+	}()
+
+	heap.Init(e.chkHeap)
+	for uint64(len(e.chkHeap.rowPtrs)) > e.chkHeap.totalLimit {
+		// The number of rows we loaded may exceeds total limit, remove greatest rows by Pop.
+		heap.Pop(e.chkHeap)
+		e.chkHeap.droppedRowNum++
+	}
+
+	if err := e.executeTopNWhenNoSpillTriggered(ctx); err != nil {
+		e.resultChannel <- rowWithError{err: err}
+		return
+	}
+
+	if e.spillHelper.isSpillNeeded() {
+		if err := e.executeTopNWhenSpillTriggered(ctx); err != nil {
+			e.resultChannel <- rowWithError{err: err}
+			return
+		}
+	}
+
+	e.generateTopNResults()
+}
+
+// Return true when spill is triggered
+func (e *TopNExec) generateTopNResultsWhenNoSpillTriggered() bool {
+	rowPtrNum := len(e.chkHeap.rowPtrs)
+	for ; e.chkHeap.idx < rowPtrNum; e.chkHeap.idx++ {
+		if e.chkHeap.idx%10 == 0 && e.spillHelper.isSpillNeeded() {
+			return true
+		}
+		e.resultChannel <- rowWithError{row: e.chkHeap.rowChunks.GetRow(e.chkHeap.rowPtrs[e.chkHeap.idx])}
+	}
+	return false
+}
+
+func (e *TopNExec) generateResultWithMultiWayMerge(offset int64, limit int64) error {
+	multiWayMerge := newMultiWayMerger(&diskSource{sortedRowsInDisk: e.spillHelper.sortedRowsInDisk}, e.lessRow)
+
+	err := multiWayMerge.init()
+	if err != nil {
+		return err
+	}
+
+	outputRowNum := int64(0)
+	for {
+		if limit != -1 && outputRowNum >= limit {
```

Comment:
when will `limit == -1`?

Reply:
> when will `limit == -1`?

It's a mistake, `limit == -1` can only appear in the previous codes, I will delete it.